### PR TITLE
feat: expose primitives for pytorch dataloaders

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

My first idea for handling iterable datasets was  #1928, but I think that PR offered tooling to the user at the wrong layer; lower-layer than what the happy-path users wanted, and higher-layer than what the advanced users needed.

This PR is another idea, which offers high-level tooling for the happy-path users, and clearly documented primitives for building data loaders that are "determined-compliant" in other cases.

Here are some meeting notes from the #ml-ag meeting on 2 Nov 2020 (abriged):

> Aaron: Accept arbitrary Pytorch DataLoaders to reduce initial porting overhead. Could also upstream some of this.
> - AH: We wanted to do it for at least IterableDatasets, but we should do this for all data loaders (while explaining caveats). Doesn’t require YogaDL necessarily (but may have caveats to behaviors). 1. It’s up to users to make sure that each GPU gets a unique data sample. 2. Pause/restart for HP search.
> - KP: question: they’ll need to do some work to get the GPU sampling to work? What does this look like from a user perspective?
>   - RB: maybe the context should offer a special sampler… (editor's note: #1928 was a failed attempt at this)
>   - AH: the other problem is the start/restart logic. We shouldn’t even try to fix that if they’re using a PT data loader
>   - RB: you’d fix that in the sampler anyway
>   - AH: our easy 1-line config change is for a DDL. if you don’t want to use it, there are these disclaimers, will take more on your side. Kind of what we do for TF dataset right now, restart logic is just not correct unless you use YogaDL but we don’t force that
> - VM: devil’s in the details, what do we do if we get a custom data loader that doesn’t work w/ Determined?
>   - AH: no, we don’t fail for TF datasets right now etc.
>   - RB: the answer is some people want it to fail, some people don’t. Like compiler warnings. We kind of need a feature like this. Document correct errors, which conditions, and allow users to request strictness depending on their use case
>   - AH: i think this is useful, but i also don’t think we should block on this
>   - VM: we need to let people know what they’re buying into; 1. If a user hasn’t read the docs and say BYODL, there’s no warning that there’s a science-compromising issue

As may be clear from the above discourse, the issue here has less to do with difficulty of implementation and more to do with correctness and being able to give assurance to the user about correctness.  That gets harder as we move away from plugin-api paradigms and into sprinkle api paradigms.  However, no users have ever complained about the pytorch-flexible-primitives API, where we made a similar transition from rigid, guaranteed-correct APIs to a "remember to make these calls" style of API.

**The conclusion that this PR takes is: "make the happy-path easy and clearly document how to make the advanced path correct"**

This "clear documentation" is at the top of `harness/determined/pytorch/samplers.py` right now. 

## Example usage:

This example uses the sampler primitives directly, as an academic exercise.

```python
class MNistTrial(PyTorchTrial):
    def build_training_data_loader(self) -> torch.utils.data.DataLoader:
        train_data = data.get_dataset(self.download_directory, train=True)

        mode = "I like pain"

        if mode == "I like easy":
            return det.pytorch.DataLoader(train_data)

        if mode == "I like pain":

            from determined.pytorch import samplers

            seed = self.context.get_trial_seed()
            rank = self.context.distributed.get_rank()
            size = self.context.distributed.get_size()
            skip = self.context.get_initial_batch()
            batch_size = self.context.get_per_slot_batch_size()

            # Shuffle-repeat-shard-batch-skip, just like Determined told me to.
            sampler = torch.utils.data.SequentialSampler(train_data)
            sampler = samplers.ReproducibleShuffleSampler(sampler, seed)
            sampler = samplers.RepeatSampler(sampler)
            sampler = samplers.DistributedSampler(sampler, num_workers=size, rank=rank)
            batch_sampler = torch.utils.data.BatchSampler(sampler, batch_size, drop_last=False)
            batch_sampler = samplers.SkipBatchSampler(batch_sampler, skip)

            return torch.utils.data.DataLoader(train_data, batch_sampler=batch_sampler)
```

## A more realistic advanced use case:

This is how @katport could have ported the detectron2 model dataloaders, with their fancy IterableDataset-based speedups.

See the original function [here](https://github.com/katport/determined/blob/DET-3158/examples/experimental/trial/detectron2/detectron2_files/data.py#L407)

```python
def build_detection_train_loader(cfg, batch_size, mapper=None):
    """
    A data loader is created by the following steps:
    1. Use the dataset names in config to query :class:`DatasetCatalog`, and obtain a list of dicts.
    2. Coordinate a random shuffle order shared among all processes (all GPUs)
    3. Each process spawn another few workers to process the dicts. Each worker will:
       * Map each metadata dict into another format to be consumed by the model.
       * Batch them by simply putting dicts into a list.
    The batched ``list[mapped_dict]`` is what this dataloader will yield.
    Args:
        cfg (CfgNode): the config
        mapper (callable): a callable which takes a sample (dict) from dataset and
            returns the format to be consumed by the model.
            By default it will be `DatasetMapper(cfg, True)`.
    Returns:
        an infinite iterator of training data
    """
    dataset_dicts = get_detection_dataset_dicts(
        cfg.DATASETS.TRAIN,
        filter_empty=cfg.DATALOADER.FILTER_EMPTY_ANNOTATIONS,
        min_keypoints=cfg.MODEL.ROI_KEYPOINT_HEAD.MIN_KEYPOINTS_PER_IMAGE
        if cfg.MODEL.KEYPOINT_ON
        else 0,
        proposal_files=cfg.DATASETS.PROPOSAL_FILES_TRAIN if cfg.MODEL.LOAD_PROPOSALS else None,
    )
    dataset = DatasetFromList(dataset_dicts, copy=False)

    if mapper is None:
        mapper = DatasetMapper(cfg, True)
    dataset = MapDataset(dataset, mapper)

    sampler_name = cfg.DATALOADER.SAMPLER_TRAIN
    print ('sampler_name: ', sampler_name)
    logger = logging.getLogger(__name__)
    logger.info("Using training sampler {}".format(sampler_name))
    # TODO avoid if-else?
    if sampler_name == "TrainingSampler":
        sampler = TrainingSampler(len(dataset))
    elif sampler_name == "RepeatFactorTrainingSampler":
        repeat_factors = RepeatFactorTrainingSampler.repeat_factors_from_category_frequency(
            dataset_dicts, cfg.DATALOADER.REPEAT_THRESHOLD
        )
        sampler = RepeatFactorTrainingSampler(repeat_factors)
    else:
        raise ValueError("Unknown training sampler: {}".format(sampler_name))

################################
# Use the new sampler primitives

    # We know:
    #  - the shuffle is handled by TrainingSampler
    #  - the repeat is handled by TrainingSampler
    #  - we still need sharding (detectron2 only shards for torch-native dtrain)
    #  - we still need skipping to support for continuing training
    #
    # We assume:
    #  - reproducibility is not required; it is enough to train on fresh data
    #    when we continue from a checkpoint.  It would be nearly impossible
    #    to repopulate the buckets of the AspectRatioGroupedDataset in a
    #    reproducible way.
    from determined.pytorch import samplers

    context = ... # You need to get this object here somehow

    # Shard before skip:
    rank = context.distributed.get_rank()
    size = context.distributed.get_size()
    sampler = samplers.DistributedSampler(sampler, num_workers=size, rank=rank)

    # Skip at the record level, since all batches are the same size in training
    batch_size = context.get_per_slot_batch_size()
    initial_batch = context.get_initial_batch()
    skip_records = batch_size * initial_batch
    sampler = det.pytorch.samplers.SkipSampler(sampler, skip=skip_records)

##############################

    return build_batch_data_loader(
        dataset,
        sampler,
        cfg.SOLVER.IMS_PER_BATCH,
        aspect_ratio_grouping=cfg.DATALOADER.ASPECT_RATIO_GROUPING,
        num_workers=cfg.DATALOADER.NUM_WORKERS,
        batch_size=batch_size
    )
```

### CODE REVIEW

File: examples/tutorials/mnist_pytorch/model_def.py [hunk 1]
```diff
@@ -61,9 +61,34 @@ def build_training_data_loader(self) -> DataLoader:
             self.data_downloaded = True
 
         train_data = data.get_dataset(self.download_directory, train=True)
-        return DataLoader(train_data, batch_size=self.context.get_per_slot_batch_size())
 
-    def build_validation_data_loader(self) -> DataLoader:
+        mode = "I like pain"
```
Comment (katport) [inline examples/tutorials/mnist_pytorch/model_def.py:65]: I want to call out that this helps porting in some cases, but this does not solve the case where a user has a get_data function that returns a dataloader and just passing that into build_data_loader. Talking to @rb-determined-ai it doesn't seem possible.

Comment (rb-determined-ai) [reply examples/tutorials/mnist_pytorch/model_def.py:65]: Right, with this change, they can do exactly that, but if it's wrong, it's wrong.  There's no way for us to modify their existing data loader to make it correct, we can only give the user to tools to take care of it themselves.

If they have a map dataset, they can still return a `det.pytorch.DataLoader` as before.

File: harness/determined/pytorch/samplers.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
+"""
+Guidelines for Reproducible Datasets.
+
+.. note::
+
+   If you are using a MapDataset and not an IterableDataset, make your life easy and use the
+   context.TrainingDataLoader() and context.ValidationDataLoader(), which are a drop-in replacement
+   for pytorch's built-in DataLoader, and Determined will handle all of these details.
```
Comment (katport) [inline harness/determined/pytorch/samplers.py:8]: add link to trainingdataloader docs

File: harness/determined/pytorch/samplers.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+
+Achieving a reproducible dataset that is able to pause and continue (sometimes called "incremental
+training") is easy if you follow a few rules.
+
+- Even if you are going to ultimate return an IterableDataset, it is best to use PyTorch's Sampler
+  class as the basis for choosing the order of which records to read.  Operations on Samplers are
+  quick and cheap, while operations on data afterwards are expensive.  For more details, see the
+  discussion at TODO:ref_yogadl_readme.  If you don't have a custom sampler, start with a simple
+  one:
```
Comment (katport) [inline harness/determined/pytorch/samplers.py:17]: "choosing the order of records"

Comment (katport) [reply harness/determined/pytorch/samplers.py:17]: "Operations on Samplers are
  quick and cheap, while operations on data afterwards are expensive.  For more details, see the
  discussion at TODO:ref_yogadl_readme. " is a bit confusing. Can we add a one line sentence that elaborate a bit more. Maybe though the TODO will help though

File: harness/determined/pytorch/samplers.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+  .. code::python
+
+     sampler = torch.utils.data.SequentialSampler(my_dataset)
+
+- **Shuffle first**: Always use a reproducible shuffle when you shuffle. Determined provides a
+  ReproducibleShuffleSampler for this purpose.  Determined also provides a
+  ReproducibleShuffleBatchSampler class, but you should prefer to shuffle at the record level
+  whenever possible, to achieve the highest-quality shuffle.
```
Comment (katport) [inline harness/determined/pytorch/samplers.py:26]: "but you should prefer to shuffle at the record level" - wait shouldnt the determined shuffler do that? Would it be better to say "if you write a custom sampler, you should shuffle at the record level ...."

File: harness/determined/pytorch/samplers.py (hunk 2 above)
Comment (katport) [reply harness/determined/pytorch/samplers.py:8]: also can we add a link of why you would use this approach: i.e. why do i need to use iterable dataset here and not the easy dropin

File: harness/determined/pytorch/samplers.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+  Determined provides a SkipBatchSampler that you can apply to your batch_sampler for this purpose.
+  There is also a SkipSampler that you can apply to your sampler, but you should prefer to skip on
+  batches unless you are confident that your dataset always yields identical size batches, where the
+  number of records to skip can be reliably calculatd from the number of batches already trained.
+
+  Always skip AFTER your repeat, so that the skip only happens once, and not on every dataset.
+
+  Always skip AFTER your shuffle, to preserve the reproducibility of the shuffle.
+
```
Comment (katport) [inline harness/determined/pytorch/samplers.py:57]: can we add a quick bullet point that reiterates above of the order i need. This way when using as a reference its like a recipe.

File: harness/determined/pytorch/samplers.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        skip,
+    ):
+        if isinstance(sampler_or_dataset, torch.utils.data.Sampler):
+            sampler = sampler_or_dataset
+        else:
+            # Create a SequentialSampler if we started with a Dataset.
+            sampler = torch.utils.data.SequentialSampler(sampler_or_dataset)
+
+        # Shuffle first.
+        if shuffle:
```
Comment (katport) [inline harness/determined/pytorch/samplers.py:80]: its a bit confusing in this example with all the else ifs. Didn't we say above that you really should use all these steps?

File: harness/determined/pytorch/samplers.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+            sampler = samplers.ReproducibleShuffleSampler(sampler, seed)
+
+        if repeat:
+            sampler = samplers.RepeatSampler(sampler)
+
+        # Always shard, and not before a repeat.
+        sampler = samplers.DistributedSampler(sampler, num_workers=num_workers, rank=rank)
+
+        # Batch before skip, because Determined counts batches, not records.
+        batch_sampler = torch.utils.data.BatchSampler(sampler, batch_size, drop_last=False)
```
Comment (katport) [inline harness/determined/pytorch/samplers.py:90]: so even if i'm using sampler I should still put it in as a batch sampler for best use with determined?

File: harness/determined/pytorch/samplers.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/samplers.py:8]: I actually think I dislike the `context.TrainingDataLoader` and `context.ValidationDataLoader()`.  I think I would feel more comfortable adding *only* the low-level API in this PR, and leave the high-level API as just the `det.pytorch.DataLoader()` (which is what we currently have)

Comment (rb-determined-ai) [reply harness/determined/pytorch/samplers.py:8]: > also can we add a link of why you would use this approach: i.e. why do i need to use iterable dataset here and not the easy dropin

Hm?  How do I know why you need to use an iterable dataset?  I feel like that's the user's concern

File: harness/determined/pytorch/samplers.py (hunk 6 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/samplers.py:80]: good point, I'll clear this code example up with a `mode` corresponding to `"training"` or `"validation"`

File: harness/determined/pytorch/samplers.py (hunk 7 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/samplers.py:90]: yes, which is explained under the `skip` bullet point

File: harness/determined/pytorch/samplers.py (hunk 4 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/samplers.py:26]: This is talking about the difference between the ReproducibleShuffleSampler and the ReproducibleShuffleBatchSampler.  I'll try to make that more obvious in the text.

File: harness/determined/pytorch/samplers.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/samplers.py:17]: it will point to yogadl.readthedocs.io, do you think that would be helpful?  (that's the README you helped edit)

File: harness/determined/pytorch/samplers.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/samplers.py:57]: Um... I'm a bit hesitant to do that because I feel like if I say:

> Remember: Shuffle-repeat-shard-batch-skip and you're good

then that simplifies things to the point of being confusing, like "why would I have to shuffle and repeat my validation dataset?"

I tried really hard to let the bolded parts of the bullets be a simple checklist:

* **Shuffle first**: ...
* **Repeat when training**: ...
* **Always shard, and not before a repeat**: ...
* **Skip when training, and always last**: ...

I'm not sure that it's very safe to simplify it any simpler than that...

File: examples/tutorials/mnist_pytorch/const.yaml [hunk 8]
```diff
@@ -15,3 +15,4 @@ searcher:
       batches: 937 #60,000 training images with batch size 64
   smaller_is_better: true
 entrypoint: model_def:MNistTrial
+max_restarts: 0
```
Comment (katport) [inline examples/tutorials/mnist_pytorch/const.yaml:18]: remove

File: harness/determined/pytorch/samplers.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+                skip=self.context.get_initial_batch(),
+            )
+
+            return torch.utils.data.DataLoader(my_dataset, batch_sampler=batch_sampler)
+"""
+from typing import Iterator
+
+import numpy as np
+import torch
+from determined_common import check
```
Comment (katport) [inline harness/determined/pytorch/samplers.py:119]: determined.common

File: harness/determined/_train_context.py [hunk 10]
```diff
@@ -176,6 +176,9 @@ def set_stop_requested(self, stop_requested: bool) -> None:
         )
         self._stop_requested = stop_requested
 
+    def get_initial_batch(self) -> int:
```
Comment (rb-determined-ai) [inline harness/determined/_train_context.py:179]: This function's existence needs more consideration.

We can only provide this information reliably when we have a lot of control over the checkpoint state (ie. not in the generic API).

Maybe it's ok to leave in the Trial context, since we do have that level of control, it almost seems like we could be a lot less magical about this by having the (very few) users who need this value do this:

```python
initial_batch = 0
last_checkpoint = generic_context.checkpointing.get_restart_checkpoint()  # this API doesn't exist yet
if last_checkpoint is not None:
    inital_batch = last_checkpoint.get_metadata("last_batch")  # this would be checkpoint metadata
                                                               # that is always saved by our
                                                               # TrialController
```

I say this is "less magical" because it makes the TrialContext a little less different than the generic api context.

File: harness/determined/pytorch/samplers.py (hunk 5 above)
Comment (katport) [reply harness/determined/pytorch/samplers.py:64]: Okay. Lets leave it then as is.

**Review by katport — APPROVED** [review]
LGTM

**Review by ioga — APPROVED** [review]
